graphql_compiler/schema_transformation/split_query.py

```diff
@@ -0,0 +1,642 @@
+)
+
+
+QueryConnection = namedtuple(
+    'QueryConnection', (
+        'sink_query_node',  # SubQueryNode
+        'source_field_out_name',
+        # str, the unique out name on the @output of the the source property field in the stitch
+        'sink_field_out_name',
+        # str, the unique out name on the @output of the the sink property field in the stitch
+    )
+)
+
+
+class SubQueryNode(object):
+    def __init__(self, query_ast):
+        """Represents one piece of a larger query, targeting one schema.
+
+        Args:
+            query_ast: Document, representing one piece of a query
+        """
+        self.query_ast = query_ast
+        self.schema_id = None  # str, identifying the schema that this query targets
+        self.parent_query_connection = None
+        # SubQueryNode or None, the query that the current query depends on
+        self.child_query_connections = []
+        # List[SubQueryNode], the queries that depend on the current query
+
+
+def split_query(query_ast, merged_schema_descriptor):
+    """Split input query AST into a tree of SubQueryNodes targeting each individual schema.
+
+    Property fields used in the stitch will be added if not already present. @output directives
+    will be added on property fields if not already present. All output names of @output
+    directives will be unique, and thus used to identify fields later down the line for adding
+    @filter directives.
+
+    Args:
+        query_ast: Document, representing a GraphQL query to split
+        merged_schema_descriptor: MergedSchemaDescriptor namedtuple, containing:
+                                  schema_ast: Document representing the merged schema
+                                  schema: GraphQLSchema representing the merged schema
+                                  type_name_to_schema_id: Dict[str, str], mapping name of each
+                                                          type to the id of the schema it came
+                                                          from
+
+    Returns:
+        Tuple[SubQueryNode, frozenset[str]]. The first element is the root of the tree of
+        QueryNodes. Each node contains an AST representing a part of the overall query,
+        targeting aspecific schema. The second element is the set of all intermediate output
+        names that are to be removed at the end
+
+    Raises:
+        - GraphQLValidationError if the query doesn't validate against the schema, contains
+          unsupported directives, or some property field occurs after a vertex field in some
+          selection
+        - SchemaStructureError if the input merged_schema_descriptor appears to be invalid
+          or inconsistent
+    """
+    check_query_is_valid_to_split(merged_schema_descriptor.schema, query_ast)
+
+    # If schema directives are correctly represented in the schema object, type_info is all
+    # that's needed to detect and address stitching fields. However, before this issue is
```

Comment:
Could be worth specifying that the issue is that GraphQL currently doesn't process schema directives correctly. I found that from reading `_get_edge_to_stitch_fields`'s comments but didn't know just from reading this. Your call if you want to include it or not